### PR TITLE
Add a `TimeDistributed` layer

### PR DESCRIPTION
**Pull request by matt-gardner**

### CODE REVIEW

**Comment by matt-gardner** [comment]
I did this earlier, when I was planning on testing the `NlpApi` a little differently.  This isn't needed for that test anymore, but we'll still need something like it eventually, so I'm creating the PR.  I also moved a few other files around a bit.

File: allennlp/modules/time_distributed.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+class TimeDistributed(torch.nn.Module):
+    """
+    Given an input shaped like ``(batch_size, time_steps, [rest])`` and a ``Module`` that takes
+    inputs like ``(batch_size, [rest])``, ``TimeDistributed`` reshapes the input to be
+    ``(batch_size * time_steps, [rest])``, applies the contained ``Module``, then reshapes it back.
+
+    Note that while the above gives shapes with ``batch_size`` first, this ``Module`` also works if
+    ``batch_size`` is second - we always just combine the first two dimensions, then split them.
+
+    This behaves identically to Keras' TimeDistributed layer.
```
Comment (DeNeutoy) [inline allennlp/modules/time_distributed.py]: Maybe remove this - no need to compare Keras, particularly as 1) Keras doesn't support having batch size second and 2) this doesn't have any concept of masking yet.

**Comment by DeNeutoy** [comment]
Ah looks like you missed some imports.
